tests/ui/rfcs/rfc-3391-result-ffi-guarantees.rs

```diff
@@ -1,3 +1,4 @@
+//@ build-pass
```

Comment:
can you add a test for your own custom result-like type here to ensure that it's not special cased to `Result`?
additionally, do we have tests that guarantee that, you know, this guarantee *actually* holds? would be good to have them :)

Reply:
The RFC does not actually extend beyond Result, actually.

Reply:
well then improper_ctypes is implemented incorrectly, because it does this for all result-like enums. the comment above the feature declaration also says result-like

Reply:
I'm happy to write more tests anyways, mind.

Reply:
well, we'd need to make sure what we actually want to guarantee before writing tests or changing the lint^^

Reply:
> well then improper_ctypes is implemented incorrectly, because it allows all result-like enums. the comment above the feature declaration also says result-like

Hm.

Reply:
i have been informed and see in the code that it also allows all option-like enums so... yolo?

Reply:
The feature description should certainly at least be correct here.

Reply:
The logic about "Option-like types" is ooooold, it exists at least Since Rust 1.5 where it got moved in [this PR](https://github.com/rust-lang/rust/pull/28580/files#diff-417fa87f2027051996a0425f1e7a71c0ff0fd049dfee6711a85f82df22574724R367).

(There's also a "FIXME: This duplicates code in trans" there, which still exists, and which I have no idea if it is still true -- it is unclear to me where that duplicate code would be and what it would do.)

Reply:
I think the original logic for recognizing Option-like types was added when the lint was originally implemented, in https://github.com/rust-lang/rust/commit/25f953437d2bc3983b03c0d17cf56de40acec45b. It was just always carried over since then.

So... I think for `Result` we don't want this, we want to specifically recognize that type. And probably for `Option`, too, but that's a different PR.